### ISSUE TITLE
fstrim service fails because no /etc/fstab

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#468]
I'm running the `next` stream and I notice after some time that the `fstrim.service` (triggered by `fstrim.timer`) fails because we don't have an `/etc/fstab`:

```
[core@localhost ~]$ systemctl status fstrim.service
● fstrim.service - Discard unused blocks on filesystems from /etc/fstab
     Loaded: loaded (/usr/lib/systemd/system/fstrim.service; static; vendor preset: disabled)
     Active: failed (Result: exit-code) since Mon 2020-04-20 00:21:40 UTC; 2 days ago
TriggeredBy: ● fstrim.timer
       Docs: man:fstrim(8)
    Process: 7825 ExecStart=/usr/sbin/fstrim --fstab --verbose --quiet (code=exited, status=32)
   Main PID: 7825 (code=exited, status=32)

Apr 20 00:21:40 localhost systemd[1]: Starting Discard unused blocks on filesystems from /etc/fstab...
Apr 20 00:21:40 localhost fstrim[7825]: fstrim: failed to parse /etc/fstab: No such file or directory
Apr 20 00:21:40 localhost systemd[1]: fstrim.service: Main process exited, code=exited, status=32/n/a
Apr 20 00:21:40 localhost systemd[1]: fstrim.service: Failed with result 'exit-code'.
Apr 20 00:21:40 localhost systemd[1]: Failed to start Discard unused blocks on filesystems from /etc/fstab.
```

We should probably at least add a `ConditionPathExists` to the systemd unit for now. A longer term question is: *"should fstrim.service be able to look at systemd mounts?"*

Version Info:

```
[core@localhost ~]$ rpm-ostree status 
State: idle
AutomaticUpdates: disabled
Deployments:
● ostree://fedora:fedora/x86_64/coreos/next
                   Version: 32.20200416.1.0 (2020-04-16T17:48:33Z)
                    Commit: 19a2e3dc4932d6043d8b437ed981d658bab019cfcde0c4cfe2b8c7d4b5df574a
              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4

  ostree://fedora:fedora/x86_64/coreos/testing
                   Version: 31.20200407.2.2 (2020-04-08T14:35:24Z)
                    Commit: b82607fd924510e0c828263d7cd38f9622420fcee8a345ed495570a36ff198fa
              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4
```

**Comment by dustymabe** [comment]
This is related to: https://fedoraproject.org/wiki/Changes/EnableFSTrimTimer

**Comment by lucab** [comment]
For reference from CL land: https://github.com/coreos/bugs/issues/2591

**Comment by cmurf** [comment]
>should fstrim.service be able to look at systemd mounts?

fstrim.service is provided by util-linux, so maybe an RFE for util-linux to see if there's a conditional way to handle the no /etc/fstab case? e.g. if no /etc/fstab then still fstrim '/' which would be the lowest common denominator you'd find in any fstab.

**Comment by cgwalters** [comment]
I kind of think this timer should be something that's done via more explicit "active userspace management" like [Stratis](https://stratis-storage.github.io/), but I understand the rationale behind the way it works today.

**Comment by lucab** [comment]
Digging into upstream history, this looks like a recent regression introduced in https://github.com/karelzak/util-linux/commit/0d73ea8bf96b036cc12fae6865158c936705238d to fix https://github.com/karelzak/util-linux/issues/673. While unfortunate, I agree with that being a better default for general purpose distributions.

I think we can still follow the plan we had for CL (making the unit conditional, and having a different unit instance with an explicit path), and later discuss whether we want to have timer-based trims in FCOS (@cmurf suggestion can fit there with a timer on `/`, I think).

**Comment by jlebon** [comment]
> so maybe an RFE for util-linux to see if there's a conditional way to handle the no /etc/fstab case? e.g. if no /etc/fstab then still fstrim '/' which would be the lowest common denominator you'd find in any fstab.

Another approach which might bring it closer in behaviour to the fstab approach is to also look for active mounts which have a corresponding non-transient systemd mount unit.

**Comment by lucab** [comment]
@jlebon does your suggestion imply linking `fstrim` against `libdbus` and teaching it how to walk the manager API, or am I completely misunderstanding it?

**Comment by jlebon** [comment]
> @jlebon does your suggestion imply linking fstrim against libdbus and teaching it how to walk the manager API, or am I completely misunderstanding it?

I was initially thinking a systemd generator which instantiates a template for each active mount unit corresponding to an actual block devices. (The idea being that since it happens in early boot, we can assume only non-temporary device mounts are present.)

But that's not going to work because it'd be running at the same time as other generators which create mount units. Maybe just a `multi-user.target` service which captures the list to somewhere in `/run` and have `fstrim.service` also pick up mounts from there?

**Comment by lucab** [comment]
Bubbled this up to util-linux upstream:
 * quick fix for silencing service unit failure: https://github.com/karelzak/util-linux/pull/1016
 * further discussion on how to provide the same feature on fstab-less systems: https://github.com/karelzak/util-linux/issues/1019

**Comment by fifofonix** [comment]
This issue has now percolated to the testing stream.  Are we expecting this to bubble to stable too in the next cycle?

**Comment by lucab** [comment]
@fifofonix are you sure about that? The fix is not yet in any upstream release and I didn't see Fedora backporting it. Where are you seen the fix in F32?

**Comment by fifofonix** [comment]
Apologies if I wasn't clear.  I'm saying the issue is no longer restricted to next.  It is now present on our test stream servers.  And I was asking if we expect to let this issue through to stable in the next release cycle, i.e. are we saying it is a benign regression we are comfortable with.

For what it is worth our servers are Active Directory-joined and have auto-mounted DFS shared storage (declared in ignition) that uses the rotating Active Directory machine credentials.

**Comment by dustymabe** [comment]
hey @fifofonix - IIUC this is a cosmetic issue (i.e., no real failure). It doesn't look good, but we have been working with upstream to get it fixed. It will most likely percolate to stable with our next round of releases, but we'll hopefully have it corrected soon.

**Comment by dustymabe** [comment]
@lucab - now that the upstream discussion has settled down a bit, what are our next steps here? Wait for new upstream release? Incorporate upstream changes in FCOS now?

**Comment by lucab** [comment]
This is fixed in upstream version 2.36 (which is only in F33 right now).

**Comment by fifofonix** [comment]
This was fixed but seems to have crept back in some how to 'next'.

**Comment by dustymabe** [comment]
hey @fifofonix. It wasn't ever fixed in Fedora 32 (where all of our streams currently are), but should be once we switch over to f33 as part of #611.

**Comment by fifofonix** [comment]
Thanks @dustymabe for clarifying.  I had thought this affected all servers, all the time, and when I noticed on at least one server that the unit wasn't reporting as failed I wrongly concluded there had been a fix.  We will wait!

**Comment by dustymabe** [comment]
The failure is now gone on Fedora 33. There is [one more fixup](https://github.com/karelzak/util-linux/pull/1163) that's need to appropriately enable the `fstrim.timer` to periodically run again, but we don't need to wait for that to close out this issue, which is related to the unit failure.

The fix for this went into next stream release `33.20201006.1.0`. Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into testing stream release `33.20201116.2.0`. Please try out the new release and report issues.

**Comment by dustymabe** [comment]
The fix for this went into stable stream release `33.20201201.3.0`.

**Comment by lucab** [comment]
For reference, the timer unit got fixed too in `util-linux 2.36.1`. That reached the `testing` stream already and will be part of the next `stable` promotion too.
